chain/chaindb.go

```diff
@@ -691,18 +696,35 @@ func (cdb *ChainDB) GetChainTree() ([]byte, error) {
 	return jsonBytes, nil
 }
 
-func (cdb *ChainDB) writeReceipts(blockHash []byte, blockNo types.BlockNo, receipts *types.Receipts) {
+func (cdb *ChainDB) writeReceiptsAndOperations(block *types.Block, receipts *types.Receipts, internalOps string) {
+	hasReceipts := len(receipts.Get()) != 0
+	hasInternalOps := len(internalOps) != 0
+
+	if !hasReceipts && !hasInternalOps {
+		return
+	}
+
 	dbTx := cdb.store.NewTx()
 	defer dbTx.Discard()
 
-	val, _ := gob.Encode(receipts)
-	dbTx.Set(dbkey.Receipts(blockHash, blockNo), val)
+	blockHash := block.BlockHash()
+	blockNo := block.BlockNo()
+
+	if hasReceipts {
+		val, _ := gob.Encode(receipts)
+		dbTx.Set(dbkey.Receipts(blockHash, blockNo), val)
+	}
+
+	if hasInternalOps {
```

Comment:
It looks like that adding new field into database. Does this PR needs to reconstruct stateDB?

Reply:
As I described above, it saves all internal operations as a JSON string into the database, one record for each block.

It does not affect the state db, it is stored on the chain database

The reason to store per block instead of per txn, is to decrease I/O

When the indexer sees that a block has transactions, it must send a request to the node to retrieve the list of internal operations

Reply:
Oh yes, the nodes that will have this feature enabled need to be sync from block 1 again, because it needs to re-execute all the txns to extract the internal operations

I suspect that this feature will be enabled only on RPC watcher nodes, not on GP nodes. But this needs to be decided